src/machine_charm.py

```diff
@@ -68,6 +73,51 @@ def _read_write_endpoint(self) -> str:
     def _read_only_endpoint(self) -> str:
         return f'file://{self._container.path("/run/mysqlrouter/mysqlro.sock")}'
 
+    @property
+    def _exposed_read_write_endpoint(self) -> str:
+        return f"{self.host_address}:{self._READ_WRITE_PORT}"
+
+    @property
+    def _exposed_read_only_endpoint(self) -> str:
+        return f"{self.host_address}:{self._READ_ONLY_PORT}"
+
+    @property
+    def _tls_certificate_saved(self) -> bool:
+        """Whether a TLS certificate is available to use"""
+        return self.tls.certificate_saved
+
+    @property
+    def _tls_key(self) -> typing.Optional[str]:
+        """Custom TLS key"""
+        return self.tls.key
+
+    @property
+    def _tls_certificate(self) -> typing.Optional[str]:
+        """Custom TLS certificate"""
+        return self.tls.certificate
+
+    @property
+    def _tls_certificate_authority(self) -> typing.Optional[str]:
+        return self.tls.certificate_authority
+
+    def is_exposed(self, relation=None) -> typing.Optional[bool]:
+        return self._database_provides.is_exposed
```

Comment:
property `is_exposed` is used in the tls and machine_workload files, thus exposing a property to avoid doing `self._charm._database_provides.is_exposed` in those files